### PR TITLE
Optimize cargo locate-project --workspace

### LINKED ISSUE #15107: `cargo locate-project --workspace` loads all the workspace
### Problem

`cargo locate-project --workspace` loads all the workspace instead of just returning the path of the workspace root.

### Steps

`cargo locate-project --workspace`

### Possible Solution(s)

Instead of doing 

`src/bin/cargo/commands/locate_project.rs`
```rs
WhatToFind::Workspace => {
    workspace = args.workspace(gctx)?;
    workspace.root_manifest()
}
```
I think that 
```rs
WhatToFind::Workspace => {
    let root = args.root_manifest(gctx)?;
    let mut ws = Workspace::new_default(root.to_path_buf(), gctx);
    ws.target_dir = gctx.target_dir()?;
    ws.find_root(manifest_path)?
}
```
or something like that would be better. Because when in a big workspace with a lot of members in a workspace, it will load some unused data.

### PR DESCRIPTION
**Pull request by moabo3li**
### Summary

Optimize cargo locate-project --workspace by adding a fast path that avoids full workspace loading when possible. 

### Problem

Previously, ```cargo locate-project --workspace``` always performed full workspace loading, which includes:
Parsing all member manifests
Resolving path dependencies
Validating workspace configuration
This is expensive for large workspaces when all we need is the workspace root path.

### Solution

Introduce a new function ```find_workspace_root_with_membership_check()``` that provides a fast path by only reading the minimal set of manifests needed to determine the workspace root.

### When Fast Path is Taken 
The fast path succeeds when both conditions are met:

- The package is explicitly listed in ```workspace.members``` (including glob patterns)
- The package is not excluded via workspace.exclude

Specific scenarios:
- Workspace root manifest: Returns immediately (the manifest is its own workspace root)
- Explicit package.workspace pointer: Reads only the target workspace manifest, verifies membership
- No explicit pointer: Walks up parent directories, checking each Cargo.toml for membership

### When Slow Path is Taken

Falls back to full workspace loading when:

- Package is a member via path dependencies (implicit membership)
- Workspace has default-members defined (requires validation)
- Package is not found in explicit members list
- Any error occurs during fast path (ensures correct error messages)

Fixes #15107

### CODE REVIEW

File: src/bin/cargo/commands/locate_project.rs
Comment (weihanglo) [inline src/bin/cargo/commands/locate_project.rs:1]: Thanks!

Would you mind adding a test first to demonstrate the current behavior that loads and verifies all members, the add the fix showing the behavior change?

Comment (moabo3li) [reply src/bin/cargo/commands/locate_project.rs:1]: Done

File: tests/testsuite/locate_project.rs
Comment (weihanglo) [inline tests/testsuite/locate_project.rs:1]: We might also want to add some extra failing cases:

* Run from a nested directory of a package. The package specifies `package.workspace` to the other workspace but the workspace in its parent directory
* Run from a nested directory of a package. The outer workspace manifest doesn't have that package as a member.

The current implementation may forget to take this into account <https://github.com/rust-lang/cargo/issues/15107#issuecomment-3604805703>.

Comment (moabo3li) [reply tests/testsuite/locate_project.rs:1]: Done

File: tests/testsuite/locate_project.rs [hunk 1]
```diff
@@ -154,3 +154,37 @@ fn workspace_missing_member() {
...
+            "#,
+        )
+        .file(
+            "nested/Cargo.toml",
+            r#"
+                [package]
+                name = "nested"
+                version = "0.0.0"
+                
+                [package.workspace]
```
Comment (weihanglo) [inline tests/testsuite/locate_project.rs]: This test setup is wrong.

`package.workspace` should be a valid workspace path 

https://doc.rust-lang.org/cargo/reference/manifest.html#the-workspace-field

File: tests/testsuite/locate_project.rs [hunk 2]
```diff
@@ -126,3 +126,111 @@ fn workspace() {
...
+        )
+        .file("not-member/src/lib.rs", "")
+        .build();
+
+    p.cargo("locate-project --workspace")
+        .cwd("not-member")
+        .with_stdout_data(
+            str![[r#"
+{
+  "root": "[ROOT]/foo/Cargo.toml"
```
Comment (weihanglo) [inline tests/testsuite/locate_project.rs:704]: Shouldn't this point to the non-member package?

File: tests/testsuite/locate_project.rs [hunk 3]
```diff
@@ -126,3 +126,111 @@ fn workspace() {
...
+        )
+        .file("nested/src/lib.rs", "")
+        .build();
+
+    p.cargo("locate-project --workspace")
+        .cwd("nested")
+        .with_stdout_data(
+            str![[r#"
+{
+  "root": "[ROOT]/foo/Cargo.toml"
```
Comment (weihanglo) [inline tests/testsuite/locate_project.rs:191]: this should point to the nested package, right?

File: tests/testsuite/locate_project.rs (hunk 2 above)
Comment (moabo3li) [reply tests/testsuite/locate_project.rs:704]: With find_workspace_root, it returns the parent workspace root, even though the package is not a member

before change it should fails with error "current package believes it's in a workspace when it's not"

File: tests/testsuite/locate_project.rs (hunk 1 above)
Comment (moabo3li) [reply tests/testsuite/locate_project.rs]: Done

File: tests/testsuite/locate_project.rs (hunk 2 above)
Comment (weihanglo) [reply tests/testsuite/locate_project.rs:704]: > With find_workspace_root, it returns the parent workspace root, even though the package is not a member

Is this the desired behavior?

Please see https://github.com/rust-lang/cargo/issues/15107#issuecomment-3604805703 that we still want a minimal verification

File: tests/testsuite/locate_project.rs (hunk 1 above)
Comment (weihanglo) [reply tests/testsuite/locate_project.rs]: Done in what way?

I expected that point to a valid workspace rootz and that affects what this command returns

Comment (moabo3li) [reply tests/testsuite/locate_project.rs]: I’m sorry if I’m missing something, but I don’t understand why this is not considered valid.

nested/Cargo.toml has workspace = "..", which resolves to the parent directory.
The parent Cargo.toml contains a [workspace] section with members = ["nested"], which is a valid workspace root.

Therefore, workspace = ".." points to a valid workspace path according to the documentation.

Is there something I’m misunderstanding?

In this test:

workspace = ".." points to the parent directory

The parent workspace lists ["nested"] in members, which includes this package

Given this, the cargo metadata command returns [ROOT]/foo/Cargo.toml (the parent workspace root).

Comment (weihanglo) [reply tests/testsuite/locate_project.rs]: I was reading the wrong diff. Sorry.

https://github.com/rust-lang/cargo/pull/16423#discussion_r2637168104

In that comment I was looking for failure cases, where `package.workspace` points to some other workspaces, not the parent one. With that we can validate when workspace manifest and package manifest disagree with each other, and should return the workspace the package manifest belongs to.

This test itself is nice to have to validate the `package.workspace` working, but not the case I was looking for.

Comment (weihanglo) [reply tests/testsuite/locate_project.rs]: BTW, by failing I didn't meant they really need to fail. They are edge cases where Cargo might want to point to the right workspace rather than the wrong workspace.

Comment (moabo3li) [reply tests/testsuite/locate_project.rs]: I added these tests, plus one more to validate cases like nested/*. If the tests and solution look good, I’ll add them along with the commits made before the fix, as usual.

File: tests/testsuite/locate_project.rs (hunk 3 above)
Comment (moabo3li) [reply tests/testsuite/locate_project.rs:191]: Yes, and since its parent is `foo/cargo`, either because the `workspace = ".."` or because it is listed in `members = ["nested"]` , it will return the parent.
So I think the test is correct. Is there anything I’m missing?

File: tests/testsuite/locate_project.rs (hunk 2 above)
Comment (moabo3li) [reply tests/testsuite/locate_project.rs:704]: Done

File: tests/testsuite/locate_project.rs (hunk 3 above)
Comment (weihanglo) [reply tests/testsuite/locate_project.rs:191]: That wasn't the test I expected, as I said in <https://github.com/rust-lang/cargo/pull/16423#discussion_r2637991885>. Since you've added more. This is considered resolved. Thanks!

File: tests/testsuite/locate_project.rs [hunk 4]
```diff
@@ -126,3 +126,238 @@ fn workspace() {
...
+        )
+        .file("inner-workspace/pkg/src/lib.rs", "")
+        .build();
+
+    p.cargo("locate-project --workspace")
+        .cwd("inner-workspace/pkg")
+        .with_stdout_data(
+            str![[r#"
+{
+  "root": "[ROOT]/foo/inner-workspace/Cargo.toml"
```
Comment (weihanglo) [inline tests/testsuite/locate_project.rs]: This test doesn't make much sense to me. The `inner-workspace/pkg/Cargo.toml` will always point to the `inner-workspace/Cargo.toml` because it is the immediate parent directory that contain a workspace manifest, regardless of having a outer workspace manifest or not. 

To avoid the parent directory probiding in your code path, you might want to change `workspace = ".."` to point to a workspace manifest that is in a sibiling directory of the package manifest.

File: src/cargo/core/workspace.rs [hunk 5]
```diff
@@ -2059,6 +2059,31 @@ impl WorkspaceRootConfig {
...
+    /// - No explicit `members` list is specified (implicit membership), or
+    /// - The path matches one of the `members` patterns
+    fn is_path_member(&self, manifest_path: &Path) -> bool {
+        let root_manifest = self.root_dir.join("Cargo.toml");
+        if manifest_path == root_manifest {
+            return true;
+        }
+        match self.members {
+            Some(ref members) => members.iter().any(|mem| {
+                if mem.contains('*') {
```
Comment (weihanglo) [inline src/cargo/core/workspace.rs]: This is not the correct way to handle glob syntax, see this function for the right approach (and we might want reuse it or refactor to reuse)

https://github.com/rust-lang/cargo/blob/1f5feecdf3557f60715fedfe1dd9e083db4da04f/src/cargo/core/workspace.rs?plain=1#L2066-L2101

File: src/cargo/core/workspace.rs [hunk 6]
```diff
@@ -2059,6 +2059,31 @@ impl WorkspaceRootConfig {
         !explicit_member && excluded
     }
 
+    /// Checks if the path could be a member of this workspace.
+    ///
+    /// Returns `true` if:
+    /// - The path is the workspace root manifest itself, or
+    /// - No explicit `members` list is specified (implicit membership), or
+    /// - The path matches one of the `members` patterns
+    fn is_path_member(&self, manifest_path: &Path) -> bool {
```
Comment (weihanglo) [inline src/cargo/core/workspace.rs]: I think this is missing an important check: path dependencies.

@epage what is your though on this, as it may provide a false result?

File: src/cargo/core/workspace.rs (hunk 5 above)
Comment (moabo3li) [reply src/cargo/core/workspace.rs]: Done

File: tests/testsuite/locate_project.rs (hunk 4 above)
Comment (moabo3li) [reply tests/testsuite/locate_project.rs]: Done

File: src/cargo/core/workspace.rs (hunk 6 above)
Comment (moabo3li) [reply src/cargo/core/workspace.rs]: Has there been any update or thoughts regarding the path dependencies check ?

Comment (weihanglo) [reply src/cargo/core/workspace.rs]: Briefly discussed this with epage. We would lean towards not having Cargo produce false results.

Path dependencies are implicit workspaces dependency if reside in the workspace directory (see [the doc](https://doc.rust-lang.org/nightly/cargo/reference/workspaces.html#the-members-and-exclude-fields) for complete definition). To determine whether a package is a path dependencies of any workspace, we need to [load the manifest file](https://github.com/rust-lang/cargo/blob/04bd5bba5d91404aac681ea7d516c0a8c5bfadbf/src/cargo/core/workspace.rs?plain=1#L826-L837), which undermines the performance improvement we want here for #15107

Ed has proposed a solution: If a package is explicitly listed as a member (in `workspace.members`), it is fine to skip loading manifest as Cargo knows it belongs to the workspace,. For package that is unknown whether it belongs to any workspace in its ancestor directories, Cargo needs to fallback to the all way (load everything) to deteremine.

Comment (moabo3li) [reply src/cargo/core/workspace.rs]: I did that and tested it to check the dependencies.

Comment (epage) [reply src/cargo/core/workspace.rs]: The one downside to this change is the error reporting will be conditioned on how the manifest is loaded.

Not saying we shouldn't do my idea but we need to be cognizant of that downside when deciding to move forward with it

Comment (moabo3li) [reply src/cargo/core/workspace.rs]: Thanks for the feedback. I’ve applied the discussed changes and tested them.

Let me know if there’s anything else you’d like me to update.

File: src/bin/cargo/commands/locate_project.rs [hunk 7]
```diff
@@ -27,15 +27,33 @@ pub struct ProjectLocation<'a> {
...
+            // Try fast path first - only works when package is explicitly listed in members
+            if let Some(ws_root) =
+                cargo::core::find_workspace_root_with_metadata(&root_manifest, gctx)?
+            {
+                workspace_root = ws_root;
+                &workspace_root
+            } else {
+                // Fallback to full workspace loading for path dependency membership.
+                // If loading fails (e.g., package is not a member of any workspace),
+                // treat the package as its own workspace root.
```
Comment (weihanglo) [inline src/bin/cargo/commands/locate_project.rs]: Why the fallback for loading failure? This sounds like a behavior change than an optimization.

File: tests/testsuite/locate_project.rs [hunk 8]
```diff
@@ -220,13 +221,184 @@ fn workspace_not_a_member() {
...
+                name = "outer-member"
+                version = "0.0.0"
+            "#,
+        )
+        .file("outer-member/src/lib.rs", "")
+        .file(
+            "sibling-workspace/Cargo.toml",
+            r#"
+                [workspace]
+                members = ["../pkg"]
```
Comment (weihanglo) [inline tests/testsuite/locate_project.rs:281]: This line is redundant, no? We already have `package.workspace = "../sibling-workspace"`. What do we want to test here?

Comment (weihanglo) [reply tests/testsuite/locate_project.rs:281]: Also, could we add these test before the fix like other tests, so the test snapshot diff will show the behavior change?

File: tests/testsuite/locate_project.rs [hunk 9]
```diff
@@ -220,13 +221,184 @@ fn workspace_not_a_member() {
         .file("not-member/src/lib.rs", "")
         .build();
 
-    //FIXME: After fix, will find workspace root without validating membership
     p.cargo("locate-project --workspace")
         .cwd("not-member")
-        .with_stderr_data(str![[r#"
-[ERROR] current package believes it's in a workspace when it's not:
```
Comment (weihanglo) [inline tests/testsuite/locate_project.rs]: To avoid this error, we should make `not-member` a workspace, which requires putting `[workspace]` in the manifest`. I think we can have both tests:

* failed with this error
* succeeded if in subdirector but having `[workspace]`.

File: tests/testsuite/locate_project.rs [hunk 10]
```diff
@@ -220,13 +221,184 @@ fn workspace_not_a_member() {
...
+                [package]
+                name = "not-member"
+                version = "0.0.0"
+            "#,
+        )
+        .file("not-member/src/lib.rs", "")
+        .build();
+
+    p.cargo("locate-project --workspace")
+        .cwd("not-member/src")
```
Comment (weihanglo) [inline tests/testsuite/locate_project.rs]: The only difference between this and `workspace_not_a_member` is that the cwd is in `not-member/src` for? Why testing this? And if we need it, we may want to just adding another `p.cargo("locate-project --workspace")` call in `workspace_not_a_member`.

File: tests/testsuite/locate_project.rs [hunk 11]
```diff
@@ -154,3 +154,38 @@ fn workspace_missing_member() {
...
 }
+
+#[cargo_test]
+fn workspace_nested_with_explicit_pointer() {
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+                [workspace]
+                members = ["nested"]
```
Comment (weihanglo) [inline tests/testsuite/locate_project.rs:170]: if we have `package.workspace = ".."` already in `nested/Cargo.toml`, do we still need `workspace.members` in the root `Cargo.toml`?

File: src/cargo/core/workspace.rs [hunk 12]
```diff
@@ -2157,6 +2186,75 @@ pub fn find_workspace_root(
     })
 }
 
+/// Finds the workspace root for a manifest, with minimal verification.
+///
+/// This is similar to `find_workspace_root`, but additionally verifies that the
+/// package and workspace agree on each other:
+/// - If the package has an explicit `package.workspace` pointer, it is trusted
+/// - Otherwise, the workspace must include the package in its `members` list
+pub fn find_workspace_root_with_metadata(
```
Comment (weihanglo) [inline src/cargo/core/workspace.rs]: Instead of reimplementing most of the probing logic, I guess it might be possible to reuse `find_workspace_root_with_loader` with a provided loader?

It might look like

```rust
pub fn find_workspace_root_with_metadata(
   …
) -> CargoResult<Option<PathBuf>> {
    find_workspace_root_with_loader(manifest_path, gctx, |self_path| {
        let source_id = SourceId::for_manifest_path(self_path)?;
        let manifest = read_manifest(self_path, source_id, gctx)?;
        match manifest.workspace_config() {
            // check explicit membership...
        }
    })

}
```

Comment (weihanglo) [inline src/cargo/core/workspace.rs]: The name `find_workspace_root_with_metadata` is not ideal. `with_metadata` is less meaningful to API user. `find_workspace_root_with_membership_check` might be a better name for the exposed API.

File: src/bin/cargo/commands/locate_project.rs (hunk 7 above)
Comment (moabo3li) [reply src/bin/cargo/commands/locate_project.rs]: Ok, I think it should be like the old one:
```Rust
workspace = args.workspace(gctx)?;
workspace.root_manifest()
```
This way, we keep the fast path and the slow path for full validation, following the old approach.

File: tests/testsuite/locate_project.rs (hunk 8 above)
Comment (moabo3li) [reply tests/testsuite/locate_project.rs:281]: Is it okay to add all the tests in one commit before the fix commit?

Comment (weihanglo) [reply tests/testsuite/locate_project.rs:281]: yes

File: tests/testsuite/locate_project.rs (hunk 10 above)
Comment (moabo3li) [reply tests/testsuite/locate_project.rs]: this test that it doesn't get confused by being one level deeper and still correctly identifies.
I`ll add it to  workspace_not_a_member and remove this test

File: tests/testsuite/locate_project.rs (hunk 8 above)
Comment (moabo3li) [reply tests/testsuite/locate_project.rs:281]: I think I forgot to remove it before. I’ll removed it now.

File: src/bin/cargo/commands/locate_project.rs (hunk 7 above)
Comment (moabo3li) [reply src/bin/cargo/commands/locate_project.rs]: Done

File: tests/testsuite/locate_project.rs (hunk 8 above)
Comment (moabo3li) [reply tests/testsuite/locate_project.rs:281]: Done

File: tests/testsuite/locate_project.rs (hunk 9 above)
Comment (moabo3li) [reply tests/testsuite/locate_project.rs]: Done

File: tests/testsuite/locate_project.rs (hunk 10 above)
Comment (moabo3li) [reply tests/testsuite/locate_project.rs]: Done

File: tests/testsuite/locate_project.rs (hunk 11 above)
Comment (moabo3li) [reply tests/testsuite/locate_project.rs:170]: Done

File: src/cargo/core/workspace.rs (hunk 12 above)
Comment (moabo3li) [reply src/cargo/core/workspace.rs]: Done

Comment (moabo3li) [reply src/cargo/core/workspace.rs]: Done

File: tests/testsuite/locate_project.rs [hunk 13]
```diff
@@ -126,3 +126,274 @@ fn workspace() {
...
+"#]]
+            .is_json(),
+        )
+        .run();
+}
+
+#[cargo_test]
+fn workspace_nested_with_explicit_pointer() {
+    let p = project()
+        .file("Cargo.toml", "")
```
Comment (weihanglo) [inline tests/testsuite/locate_project.rs]: This workspace manifest is not a valid manifest. I would assume this to fail. Having an explicit pointer in member manifest doesnt mean workspace manifest agrees with the membership.

File: tests/testsuite/locate_project.rs [hunk 14]
```diff
@@ -126,3 +126,274 @@ fn workspace() {
...
+        )
+        .file("not-member/src/lib.rs", "")
+        .build();
+
+    p.cargo("locate-project --workspace")
+        .cwd("not-member")
+        .with_status(101)
+        .with_stderr_data(str![[r#"
+[ERROR] current package believes it's in a workspace when it's not:
+...
```
Comment (weihanglo) [inline tests/testsuite/locate_project.rs]: Why are we omitting here? I feel like it is fine snapshotting the entire stderr, no?

File: tests/testsuite/locate_project.rs
Comment (weihanglo) [inline tests/testsuite/locate_project.rs]: This should be a valid workspace manifest.

File: src/cargo/core/workspace.rs [hunk 15]
```diff
@@ -2162,6 +2186,48 @@ pub fn find_workspace_root(
...
+
+    match current_manifest.workspace_config() {
+        WorkspaceConfig::Root(_) => {
+            // This manifest is a workspace root itself
+            Ok(Some(manifest_path.to_path_buf()))
+        }
+        WorkspaceConfig::Member {
+            root: Some(path_to_root),
+        } => {
+            // Has explicit `package.workspace` pointer - trust it
```
Comment (weihanglo) [inline src/cargo/core/workspace.rs]: No. We should not. We should check if the workspace manifest agrees with it.

So basically we'll always have to read at least two manifests for inspecting a member manifest.

File: src/cargo/core/workspace.rs [hunk 16]
```diff
@@ -2064,6 +2059,35 @@ impl WorkspaceRootConfig {
...
+        }
+        match self.members {
+            Some(ref members) => {
+                // Use members_paths to properly expand glob patterns
+                let Ok(expanded_members) = self.members_paths(members) else {
+                    return false;
+                };
+                expanded_members
+                    .iter()
+                    .any(|(member_path, _)| manifest_path.starts_with(member_path))
```
Comment (weihanglo) [inline src/cargo/core/workspace.rs]: Instead of `starts_with` we probably want to compare the path explicitly. Something like

```suggestion
                    .any(|(member_path, _)| manifest_path.parent() == Some(member_path))
```

Also, you might want to check whether the `member_path` is not normalized (e.g., `../crates/*` expand to `../crates/foo/Cargo.toml` without removing `..`). There are a couple of `paths::normalize_path` call in the module I am aware of.

File: src/cargo/core/workspace.rs [hunk 17]
```diff
@@ -2064,6 +2059,35 @@ impl WorkspaceRootConfig {
...
+    ///
+    /// Returns `true` ONLY if:
+    /// - The path is the workspace root manifest itself, or
+    /// - The path matches one of the explicit `members` patterns
+    ///
+    /// NOTE: This does NOT check for implicit path dependency membership.
+    /// A `false` return does NOT mean the package is definitely not a member -
+    /// it could still be a member via path dependencies. Callers should fallback
+    /// to full workspace loading when this returns `false`.
+    fn is_explicitly_listed_member(&self, manifest_path: &Path) -> bool {
```
Comment (weihanglo) [inline src/cargo/core/workspace.rs:2072]: https://github.com/rust-lang/cargo/blob/3861f60f6b58f57524c0e7aab1f5c1ad83e35409/src/cargo/core/workspace.rs?plain=1#L841-L846

I found this interesting comment. Wonder what would happen with the current implementation in these situations:

* A member is listed in `workspace.default-members` but not in `workspace.members`.
* A member is listed in both `workspace.default-members` and `workspace.exclude`
* A member is listed in both `workspace.members`  and `workspace.exclude`

The new implementation should be aligned with the default `args.workspace()` behavior IMO.

File: tests/testsuite/locate_project.rs (hunk 13 above)
Comment (moabo3li) [reply tests/testsuite/locate_project.rs]: I’ll do it, but what do you mean by fail?
I mean, the test has already passed.

File: tests/testsuite/locate_project.rs (hunk 14 above)
Comment (moabo3li) [reply tests/testsuite/locate_project.rs]: Done

File: tests/testsuite/locate_project.rs (hunk 13 above)
Comment (moabo3li) [reply tests/testsuite/locate_project.rs]: Done

File: tests/testsuite/locate_project.rs
Comment (moabo3li) [reply tests/testsuite/locate_project.rs]: What do you want to mention here, since the code isn’t shown?

Comment (weihanglo) [reply tests/testsuite/locate_project.rs]: Sorry I meant   `.file("sibling-workspace/Cargo.toml", "")`. If the workspace manifest isn't valid, the `cargo locate-project` should fail.

File: tests/testsuite/locate_project.rs (hunk 13 above)
Comment (weihanglo) [reply tests/testsuite/locate_project.rs]: This <https://github.com/rust-lang/cargo/pull/16423#discussion_r2662880917>.

Invalid workspace manifest should fail the invocation of `cargo locate-project` but I believed it wasn't the intent of this test.

File: src/cargo/core/workspace.rs (hunk 17 above)
Comment (moabo3li) [reply src/cargo/core/workspace.rs:2072]: I checked whether the manifest has a default member. If it does, I run the full check using `args.workspace()`.

File: src/cargo/core/workspace.rs (hunk 16 above)
Comment (moabo3li) [reply src/cargo/core/workspace.rs]: Done

File: src/cargo/core/workspace.rs (hunk 15 above)
Comment (moabo3li) [reply src/cargo/core/workspace.rs]: Done

File: tests/testsuite/locate_project.rs
Comment (moabo3li) [reply tests/testsuite/locate_project.rs]: Done

**Comment by moabo3li** [comment]
I think after we finalize this, we may need to update the PR description to describe the new flow and when the fast or slow path is taken.
Do you think the current description is fine?

File: tests/testsuite/locate_project.rs (hunk 11 above)
Comment (weihanglo) [reply tests/testsuite/locate_project.rs:170]: I meant we should probably remove these two lines, as there is already an explicit pointer, and we want to test `package.workspace` works even with out `workspace.members` specifying the member crate. Do I understand the intent of this test correctly?

```

                [workspace]
                members = ["nested"]
```

File: tests/testsuite/locate_project.rs [hunk 18]
```diff
@@ -126,3 +126,679 @@ fn workspace() {
...
+                members = ["../pkg"]
+            "#,
+        )
+        .file(
+            "pkg/Cargo.toml",
+            r#"
+                [package]
+                name = "pkg"
+                version = "0.0.0"
+                workspace = "../sibling-workspace"
```
Comment (weihanglo) [inline tests/testsuite/locate_project.rs:294]: We probably just need either `workspace.members` or `package.workspacer`. We don't need both.

File: tests/testsuite/locate_project.rs [hunk 19]
```diff
@@ -126,3 +126,679 @@ fn workspace() {
...
+        )
+        .file("pkg/src/lib.rs", "")
+        .build();
+
+    p.cargo("locate-project --workspace")
+        .cwd("pkg")
+        .with_stdout_data(
+            str![[r#"
+{
+  "root": "[ROOT]/foo/pkg/Cargo.toml"
```
Comment (weihanglo) [inline tests/testsuite/locate_project.rs:576]: This looks like a bug to me 😞. I ran `cargo build` within `foo/pkg/` directoyr, and the lockfile was created at `foo/pkg/Cargo.lock`. I ran `cargo metadata` at `foo/` and it failed with this message.

```
cargo metadata
warning: please specify `--format-version` flag explicitly to avoid compatibility problems
error: manifest path `/Users/whlo/dev/cargo-dev/target/tmp/cit/t0/foo` contains no package: The manifest is virtual, and the workspace has no members.
```

I think `cargo metadata` does it right, `cargo locate-project` is buggy.

Maybe worth fixing it first? before we land the optimization?

File: tests/testsuite/locate_project.rs [hunk 20]
```diff
@@ -126,3 +126,679 @@ fn workspace() {
...
+                members = ["../crates/*"]
+            "#,
+        )
+        .file(
+            "crates/foo/Cargo.toml",
+            r#"
+                [package]
+                name = "foo"
+                version = "0.0.0"
+                workspace = "../../workspace"
```
Comment (weihanglo) [inline tests/testsuite/locate_project.rs:634]: Ditto. We don't need both `workspace.members` and `package.workspace` co-exist unless we are specifically testing their disagreement with each other's value.

**Review by weihanglo — COMMENTED** [review]
I feel like we should land a fix for the `exclude` and `members` conflict issue before doing the optimization, to give us a good baseline to compare with.

If you’re willing, could you summarize the behavior and interaction between them in a table?  

Maybe we can compare the results of  `cargo metadata | jq -r .workspace_root` and  `cargo metadata | jq -r .workspace_members[]`,   and see whether the current `cargo locate-project` behaves correctly.

(Sorry this issue wasn't immediately clear that it is going to be this gnarly.)


*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/cargo/16423/6d1bd93c47f059ec1344cb31e68a2fb284cbc6b1..8a090fc485f1621f232f2408aa47f292e78a986e)*

File: tests/testsuite/locate_project.rs (hunk 19 above)
Comment (weihanglo) [reply tests/testsuite/locate_project.rs:576]: Not a blocker.

File: tests/testsuite/locate_project.rs (hunk 11 above)
Comment (moabo3li) [reply tests/testsuite/locate_project.rs:170]: I tested this before the fix commit. If we just remove members = ["nested"] without adding anything else, the test fails because stable cargo requires bilateral agreement for workspace membership.
With only an empty [workspace] table and no path dependency:
```
error: current package believes it's in a workspace when it's not:
current:   /home/abo-ali/cargo/target/tmp/cit/t14/foo/nested/Cargo.toml
workspace: /home/abo-ali/cargo/target/tmp/cit/t14/foo/Cargo.toml

this may be fixable by ensuring that this crate is depended on by the workspace root: /home/abo-ali/cargo/target/tmp/cit/t14/foo/Cargo.toml
Alternatively, to keep it out of the workspace, add the package to the `workspace.exclude` array, or add an empty `[workspace]` table to the package's manifest.
```
i think if i will think for another way to check if the explicit pointer or not is to use 
```
[dependencies]
nested = { path = "nested" }
```
This way we're testing that package.workspace works with implicit membership with path dependencies rather than explicit members listing.

File: tests/testsuite/locate_project.rs (hunk 18 above)
Comment (moabo3li) [reply tests/testsuite/locate_project.rs:294]: For this test we actually need both:
for ```workspace = "../sibling-workspace"```
Without this, Cargo walks up and finds the outer workspace first (which is wrong)
```
error: current package believes it's in a workspace when it's not:
current:   /home/abo-ali/cargo/target/tmp/cit/t19/foo/pkg/Cargo.toml
workspace: /home/abo-ali/cargo/target/tmp/cit/t19/foo/Cargo.toml

this may be fixable by adding `pkg` to the `workspace.members` array of the manifest located at: /home/abo-ali/cargo/target/tmp/cit/t19/foo/Cargo.toml
Alternatively, to keep it out of the workspace, add the package to the `workspace.exclude` array, or add an empty `[workspace]` table to the package's manifest.
```
and for ```members = ["../pkg"]``` the same thing here 
https://github.com/rust-lang/cargo/pull/16423#discussion_r2691533954

```
error: current package believes it's in a workspace when it's not:
current:   /home/abo-ali/cargo/target/tmp/cit/t19/foo/pkg/Cargo.toml
workspace: /home/abo-ali/cargo/target/tmp/cit/t19/foo/sibling-workspace/Cargo.toml

this may be fixable by adding a member to the `workspace.members` array of the manifest located at: /home/abo-ali/cargo/target/tmp/cit/t19/foo/sibling-workspace/Cargo.toml
Alternatively, to keep it out of the workspace, add the package to the `workspace.exclude` array, or add an empty `[workspace]` table to the package's manifest.
```

File: tests/testsuite/locate_project.rs (hunk 20 above)
Comment (moabo3li) [reply tests/testsuite/locate_project.rs:634]: This case is similar to [workspace_pointer_to_sibling_workspace](https://github.com/rust-lang/cargo/pull/16423#discussion_r2691581837). Both are needed because the packages are outside the workspace directory:
Without
```workspace = "../../workspace"``` 
 Cargo doesn't find the workspace (returns standalone)
 and 
Without 
```members = ["../crates/*"]``` 
fails
with 
```
error: current package believes it's in a workspace when it's not:
current:   /home/abo-ali/cargo/target/tmp/cit/t8/foo/crates/foo/Cargo.toml
workspace: /home/abo-ali/cargo/target/tmp/cit/t8/foo/workspace/Cargo.toml

this may be fixable by adding a member to the `workspace.members` array of the manifest located at: /home/abo-ali/cargo/target/tmp/cit/t8/foo/workspace/Cargo.toml
Alternatively, to keep it out of the workspace, add the package to the `workspace.exclude` array, or add an empty `[workspace]` table to the package's manifest.
```

Comment (weihanglo) [reply tests/testsuite/locate_project.rs:634]: Understandable. Thanks. I think we can merge this for now. Thanks for working with us!

Comment (moabo3li) [reply tests/testsuite/locate_project.rs:634]: Great, I’ll update the PR description now.
As for these, I’ll look at them after the PR gets merged.

> * [#16423 (review)](https://github.com/rust-lang/cargo/pull/16423#pullrequestreview-3652446960)
> * [#16423 (comment)](https://github.com/rust-lang/cargo/pull/16423#discussion_r2683698072)

and sorry for the delay, I was in the middle of my final exams.

Comment (moabo3li) [reply tests/testsuite/locate_project.rs:634]: I updated the PR description.

**Review by weihanglo — APPROVED** [review]
Thanks again!

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/cargo/16423/6d1bd93c47f059ec1344cb31e68a2fb284cbc6b1..8a090fc485f1621f232f2408aa47f292e78a986e)*
